### PR TITLE
SRF Dropout: regularize surface refinement heads for OOD

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model currently has **zero dropout anywhere** in the architecture. The Surface Refinement (SRF) heads are the final output layers — they take backbone hidden states and produce pressure corrections on the surface. If these heads overfit to the exact training distribution, they'll produce inaccurate corrections for OOD inputs.

**Adding dropout (p=0.1) to the SRF MLP layers** provides stochastic regularization during training, forcing the SRF heads to be robust to feature dropout. At inference, dropout is disabled (deterministic predictions).

**Why SRF heads specifically (not backbone):**
1. The backbone needs full capacity to learn physics-attention routing. Backbone dropout would interfere with the slice mechanism.
2. SRF heads are 3-layer MLPs (192-dim) that produce the final surface correction. This is where output overfitting happens — the heads memorize exact pressure corrections for training configurations.
3. Selective dropout on output heads is standard in multi-task learning (e.g., detection heads in YOLO, classification heads in ViT). It regularizes the prediction without constraining feature learning.

**Physical motivation:** For OOD inputs (NACA6416 tandem), the backbone produces slightly different hidden states than for training geometries. The SRF heads need to generalize their correction function to these unseen states. Dropout forces the heads to not rely on any single feature dimension, making the correction more robust.

**Confidence:** Medium. Dropout is well-established but has been less effective on physics models than on classification. The key question is whether p=0.1 is enough (or too much) for these small MLPs.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flag

```python
srf_dropout: float = 0.0    # dropout probability in SRF MLP layers (0 = disabled)
```

### Step 2: Add dropout layers to SRF heads

Find where the Surface Refinement MLP layers are constructed. They should be sequential Linear → Activation → Linear chains. Add `nn.Dropout(cfg.srf_dropout)` after each activation:

```python
# In the SRF head construction (adjust based on actual code structure):
if cfg.srf_dropout > 0:
    layers = []
    for i in range(n_layers):
        layers.append(nn.Linear(in_dim, out_dim))
        layers.append(nn.GELU())  # or whatever activation
        layers.append(nn.Dropout(cfg.srf_dropout))
    layers.append(nn.Linear(out_dim, output_dim))  # final projection, no dropout
    srf_mlp = nn.Sequential(*layers)
```

**IMPORTANT:** Do NOT add dropout after the final Linear layer — only between hidden layers. The output projection should be deterministic.

Apply to BOTH the fore-foil SRF head and the aft-foil SRF head (if separate).

### Step 3: Verify dropout is training-only

`nn.Dropout` automatically disables during `model.eval()`. Verify that validation/visualization code calls `model.eval()` (or `ema_model.eval()`). No manual handling needed.

### Step 4: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/srf-dropout-s42" \
  --wandb_group "round19/srf-dropout" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --srf_dropout 0.1

# Seed 73 — identical but --seed 73 --wandb_name "alphonse/srf-dropout-s73"
```

### Step 5: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

Focus on OOD metrics (p_oodc, p_re) — these are the expected beneficiaries of regularization.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

W&B: `hgml7i2r` (s42), `qic03vrg` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Summary

SRF Dropout (p=0.1) showed a **mixed result** — p_oodc improved marginally but p_tan regressed substantially. Net result is a regression vs baseline.

### Metrics Table

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | vs Baseline |
|--------|----------|---------|---------|------------|-------------|
| **p_in**  | 11.979 | 12.107 | 12.243 | **12.175** | +0.196 (+1.6%, worse) |
| **p_oodc** | 7.643 | 7.687 | 7.573 | **7.630** | **-0.013 (-0.2%, better!)** |
| **p_tan** | 28.341 | 31.067 | 30.614 | **30.841** | +2.500 (+8.8%, much worse) |
| **p_re** | 6.300 | 6.587 | 6.478 | **6.533** | +0.233 (+3.7%, worse) |

### Run Details

| | Seed 42 | Seed 73 |
|--|---------|---------|
| W&B run | `nq2e1w8v` | `d46kd21a` |
| Best epoch | 148 / 500 (wall-clock) | 147 / 500 (wall-clock) |
| VRAM | 46.2 GB | 46.3 GB |
| Epoch time | 72-73s | 73-74s |

### Command

```bash
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/srf-dropout-s42" \
  --wandb_group "round19/srf-dropout" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --srf_dropout 0.1
# Seed 73 identical with --seed 73 --wandb_name "alphonse/srf-dropout-s73"
```

### What Happened

1. **p_oodc improved slightly (-0.013)**: The SRF dropout did provide a small regularization benefit for OOD conditions, as expected. However, the effect is negligible (~0.2%) and could be within noise.

2. **p_tan regressed severely (+2.500 = +8.8%)**: This is the main failure. The aft-foil SRF head is responsible for correcting pressure on aft-foil nodes, which is the key challenge for tandem transfer (OOD geometry). By dropout-regularizing this head, we're making it LESS precise at the tandem correction task. The SRF heads' job is to make fine-grained corrections — dropout forces them to rely on partial feature subsets, degrading this precision exactly where it matters most.

3. **p_re regressed (+0.233)**: Reynolds number OOD also worsened, suggesting the SRF heads need full capacity to generalize across Re.

4. **p_in regressed slightly (+0.196)**: Consistent with the SRF heads being less precise at the main task.

The regularization story here is backwards: the SRF heads are ALREADY well-constrained by zero initialization and LayerNorm. Adding dropout on top creates an underfitting problem rather than solving overfitting. The SRF heads' job is precise correction; dropout introduces variance that directly hurts correction accuracy.

### Suggested Follow-ups

- **Lower dropout (p=0.05 or p=0.02)**: The severe p_tan regression suggests p=0.1 is too aggressive. A much smaller value might preserve aft-foil SRF precision while providing mild regularization.
- **Dropout on fore-SRF only, not aft-SRF**: The fore-SRF is for in-distribution predictions; the aft-SRF is specifically for OOD tandem transfer. Selectively applying dropout only to the fore-SRF head might help p_oodc/p_in without hurting p_tan.
- **Backbone-side regularization instead**: The SRF heads are output heads — regularizing them doesn't help backbone OOD generalization. Applying mild dropout to the Transolver attention scores or slice features might be more effective.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2245 because ADVISOR: Closing — SRF Dropout is counterproductive.

Results (2-seed avg vs baseline):
- p_in: 12.175 vs 11.979 (+1.6%)
- p_oodc: 7.630 vs 7.643 (-0.2% — within noise)
- p_tan: 30.841 vs 28.341 (+8.8% — severe regression)
- p_re: 6.533 vs 6.300 (+3.7%)

Good analysis in your report. The SRF heads are output-refinement modules that need full precision — they're already well-constrained by zero initialization + LayerNorm. Dropout creates underfitting in the correction pathway rather than addressing backbone overfitting.

Key takeaway: regularizing output heads doesn't help OOD generalization because the generalization bottleneck is in the backbone, not the heads. Filed as dead end..
